streamer/src/nonblocking/connection_rate_limiter.rs

```diff
@@ -1,156 +1,85 @@
-    limiter: DefaultKeyedRateLimiter<IpAddr>,
+    limiter: KeyedRateLimiter<IpAddr>,
 }
 
+/// The threshold of the size of the connection rate limiter map. When
+/// the map size is above this, we will trigger a cleanup of older
+/// entries used by past requests.
+const CONNECTION_RATE_LIMITER_CLEANUP_SIZE_THRESHOLD: usize = 100_000;
+
 impl ConnectionRateLimiter {
     /// Create a new rate limiter per IpAddr. The rate is specified as the count per minute to allow for
-    /// less frequent connections.
+    /// less frequent connections. Higher limit also allows higher bursts.
     pub fn new(limit_per_minute: u64) -> Self {
-        let quota =
-            Quota::per_minute(NonZeroU32::new(u32::try_from(limit_per_minute).unwrap()).unwrap());
         Self {
-            limiter: DefaultKeyedRateLimiter::keyed(quota),
+            limiter: KeyedRateLimiter::new(
+                CONNECTION_RATE_LIMITER_CLEANUP_SIZE_THRESHOLD,
+                TokenBucket::new(
+                    limit_per_minute,
+                    limit_per_minute,
+                    limit_per_minute as f64 / 60.0,
+                ),
+                8,
+            ),
         }
     }
 
     /// Check if the connection from the said `ip` is allowed.
+    /// Here we assume that only IPs with actual confirmed connections are stored in it,
+    /// since we should only modify server state once source IP is verifired
     pub fn is_allowed(&self, ip: &IpAddr) -> bool {
-        // Acquire a permit from the rate limiter for the given IP address
-        if self.limiter.check_key(ip).is_ok() {
-            debug!("Request from IP {ip:?} allowed");
-            true // Request allowed
-        } else {
-            debug!("Request from IP {ip:?} blocked");
-            false // Request blocked
-        }
-    }
-
-    /// retain only keys whose rate-limiting start date is within the rate-limiting interval.
-    /// Otherwise drop them as inactive
-    pub fn retain_recent(&self) {
-        self.limiter.retain_recent()
-    }
-
-    /// Returns the number of "live" keys in the rate limiter.
-    pub fn len(&self) -> usize {
-        self.limiter.len()
-    }
-
-    /// Returns `true` if the rate limiter has no keys in it.
-    pub fn is_empty(&self) -> bool {
-        self.limiter.is_empty()
-    }
-}
-
-/// Connection rate limiter for enforcing connection rates from
-/// all clients.
-pub struct TotalConnectionRateLimiter {
-    limiter: DefaultDirectRateLimiter,
-}
-
-impl TotalConnectionRateLimiter {
-    /// Create a new rate limiter. The rate is specified as the count per second.
-    pub fn new(limit_per_second: u64) -> Self {
-        let quota =
-            Quota::per_second(NonZeroU32::new(u32::try_from(limit_per_second).unwrap()).unwrap());
-        Self {
-            limiter: RateLimiter::direct(quota),
+        // Check if we have records in the rate limiter for the given IP address
+        dbg!(self.limiter.current_tokens(ip));
```

Comment:
use the debug logging function to be consistent

Reply:
And log the conditions for debug purposes.

Reply:
Ooh thanks, good catch!

Reply:
fixed in https://github.com/anza-xyz/agave/commit/dd66eae309f73e5ac3a477145a325a2e51b0558b